airbyte-integrations/bases/base-python/base_python/sdk/streams/rate_limiting.py

```diff
@@ -0,0 +1,58 @@
+import sys
+import time
+
+import backoff
+from ...logger import AirbyteLogger
+from requests import exceptions
+from .exceptions import DefaultBackoffException, CustomBackoffException
+
+TRANSIENT_EXCEPTIONS = (DefaultBackoffException, exceptions.ConnectTimeout, exceptions.ReadTimeout)
+
+# TODO inject singleton logger?
+logger = AirbyteLogger()
+
+
+def default_backoff_handler(max_tries: int, factor: int, **kwargs):
+    def log_retry_attempt(details):
+        _, exc, _ = sys.exc_info()
+        logger.info(str(exc))
+        logger.info(f"Caught retryable error after {details['tries']} tries. Waiting {details['wait']} more seconds then retrying...")
+
+    def should_giveup(exc):
+        # return true if this is covered by the other functions. Otherwise return false.
+        return exc.response is not None and 400 <= exc.response.status_code < 500
```

Comment:
use class here to be able to override this logic?

Reply:
going to kick this to a follow on task. PR is already fairly large in scope.